## Задание
---
Подкинули монету N раз. Кол-во случаев, когда выпал орёл, на 10% больше, чем кол-во случаев, когда выпала решка. При каком N мы можем сказать, что монета «нечестная» (орёл и решка выпадают с разной вероятностью)?

## Решение
---

Подойти к решению задачи можно следующим путем:  
Для каждого рассматриваемого размера выборки применить биномиальный критерий для доли, выдвинуть гипотезу и альтернативу.
- H0 (нулевая гипотеза): 
    - монета честная, то есть вероятность выпадения орла и решки одинаковы и равны 0,5
- H1 (альтернатива): 
    - альтерантива бывает двух видов, односторонняя и двусторонняя. В данном случае, так как из условия задачи следует, что орел выпадает чаще решки, можно рассматривать одностороннюю альтернативу: вероятность выпадения орла больше 0.5
- статистика Т: 
    - количество единиц в выборке (количество случаев выпадения орла)

In [18]:
import scipy.stats as st
import pandas as pd
import numpy as np
import statsmodels as sm
from scipy import stats

In [19]:
# Функция вычисления вероятности выпадения решки при броске монеты
# difference - на сколько чаще выпадает орёл, нежели решка
# p1 = 1 - p2
# p1 = 1.1 * p2
# 2.1 * p2 = 1
def count_probability(difference):
    return 1.0 / (2.0 + difference)

In [20]:
p = 1 - count_probability(0.1)  # по условию кол-во случаев, 
                                # когда выпал орёл, на 10% больше => difference = 0.1
print("Вероятность выпадения орла:", p)

Вероятность выпадения орла: 0.5238095238095238


In [21]:
# Для начала, попробуем оценить при помощи доверительного интервала, при каком объеме выборки,
# можно сделать вывод о том, что вероятность выпадения орла не равняется 0.5
for n in range(1, 5000):
    result_conf_int = sm.stats.proportion.proportion_confint(n * p, n)
    if result_conf_int[0] > 0.5:
        print("При размере выбоки, равном", n, "\nДоверительный интервал равен", result_conf_int)
        break

При размере выбоки, равном 1691 
Доверительный интервал равен (0.5000053378100074, 0.5476137098090402)


Так как доверительный интервал не содержит 0.5, можно сказать, что при данном объёме выборки можно сделать вывод о том, что вероятность выпадения орла не равняется 0,5.  
  
Однако, в данном случае, не учитывается тот факт, что статистика односторонняя.  
  
Поэтому, применим биномиальный критерий для доли:

In [22]:
for n in range(1, 5000):
    results = stats.binom_test(n * p, n, 0.5, alternative='greater')
    if results < 0.05:
        print("При достигаемом уровне значимости 0.05, размер выборки должен составлять:", n)
        print("p-value =", results)
        break

При достигаемом уровне значимости 0.05, размер выборки должен составлять: 1239
p-value = 0.049682081779629406


## ОТВЕТ
---
При уровне значимости 0.05, мы можем отвергнуть гипотезу о честности монетки в сторону односторонней альтернативы (вероятность выпадения орда больше вероятности выпадения решки) при минимальном объеме выборки 1239.  
  
Стоит отметить, что при разных уровнях занчимости, размер выборки будет различным.